functions/index.js

```diff
@@ -0,0 +1,91 @@
+//
+// This source file is part of the StudyApplication based on the Stanford Spezi Template Application project
+//
+// SPDX-FileCopyrightText: 2023 Stanford University
+//
+// SPDX-License-Identifier: MIT
+//
+
+const {onCall} = require("firebase-functions/v2/https");
+const {logger, https} = require("firebase-functions/v2");
+const {FieldValue} = require("firebase-admin/firestore");
+const admin = require("firebase-admin");
+const {beforeUserCreated} = require("firebase-functions/v2/identity");
+
+admin.initializeApp();
+
+exports.checkInvitationCode = onCall(async (request) => {
+  if (!request.auth || !request.auth.uid) {
+    throw new https.HttpsError(
+        "unauthenticated",
+        "User is not properly authenticated.",
+    );
+  }
+
+  const {invitationCode} = request.data;
+  const userId = request.auth.uid;
+  const firestore = admin.firestore();
+
+  logger.debug(`User (${userId}) -> PAWS, InvitationCode ${invitationCode}`);
+
+  try {
+    // Based on https://github.com/StanfordSpezi/SpeziStudyApplication/blob/main/functions/index.js
+    const invitationCodeRef = firestore.doc(`invitationCodes/${invitationCode}`);
+    const invitationCodeDoc = await invitationCodeRef.get();
+
+    if (!invitationCodeDoc.exists || invitationCodeDoc.data().used) {
+      throw new https.HttpsError("not-found", "Invitation code not found or already used.");
+    }
+
+    const userStudyRef = firestore.doc(`users/${userId}`);
+    const userStudyDoc = await userStudyRef.get();
+
+    if (userStudyDoc.exists) {
+      throw new https.HttpsError("already-exists", "User is already enrolled in the study.");
+    }
+
+    await firestore.runTransaction(async (transaction) => {
+      transaction.set(userStudyRef, {
+        invitationCode: invitationCode,
+        dateOfEnrollment: FieldValue.serverTimestamp(),
+      });
+
+      transaction.update(invitationCodeRef, {
+        used: true,
+        usedBy: userId,
+      });
+    });
+
+    logger.debug(`User (${userId}) successfully enrolled in study (PAWS) with invitation code: ${invitationCode}`);
+
+    return {};
+  } catch (error) {
+    logger.error(`Error processing request: ${error.message}`);
+    if (!error.code) {
+      throw new https.HttpsError("internal", "Internal server error.");
+    }
+    throw error;
+  }
+});
+
+exports.beforecreated = beforeUserCreated(async (event) => {
+  const firestore = admin.firestore();
+
+  try {
+    // Check Firestore to confirm whether an invitation code has been associated with a user.
+    const invitationQuerySnapshot = await firestore.collection("invitationCodes")
+        .where("usedBy", "==", event.data.user.uid)
+        .limit(1)
+        .get();
+
+    if (invitationQuerySnapshot.empty) {
```

Comment:
For a complete sign up to work we should also check that the user document exists and also contains an invitation code. Might even make sense to check that is the same one that was retrieved from the query snapshot to check that the elements are completely consistent as done with the function above.

Reply:
Sounds good. I have added the logic for these checks and will let you know if the overall functionality works as expected in the emulator.

Reply:
Sounds great; thank you!

Reply:
It would be great to write a UI test around this to verify the behavior in the application.

We might need to seed some data in the Firebase Setup to do that.
You can inspect how wo do that in the Spezi Study application as well: https://github.com/StanfordSpezi/SpeziStudyApplication?tab=readme-ov-file#firebase-emulator-based-development--testing

Reply:
Sure, I will give it a try. As a heads-up, the sign-up view seems to glitch out when I successfully create an account, at which point I have to close the app, open it again, and then use a newly generated invitation code to bypass that screen to log in with the credentials I already have. It could just be a quirk of the iOS simulator or Firebase emulator, but I am not sure.